File: lib/sidereon/ccsds/oem.ex
defmodule Sidereon.CCSDS.OEM do
  @moduledoc """
  Parse and encode CCSDS Orbit Ephemeris Messages (OEM).

  Supports both the **KVN** (Keyword=Value Notation) and **XML** formats per
  CCSDS 502.0-B. An OEM carries one or more segments, each a metadata block plus
  a time-ordered list of Cartesian state samples (optionally with acceleration)
  and optional covariance blocks.

  `parse/1` auto-detects the format from the first non-whitespace character: a
  leading `<` is treated as XML, anything else as KVN. Date/time fields are
  preserved as raw strings exactly as written; the message round-trips through
  the canonical struct without calendar rewriting.

  ## Examples

      {:ok, oem} = Sidereon.CCSDS.OEM.parse(kvn_string)
      [segment | _] = oem.segments
      segment.metadata.object_name
      [state | _] = segment.states
      state.position_km            # {x, y, z} in km

      # KVN output (default)
      kvn = Sidereon.CCSDS.OEM.encode(oem)

      # XML output
      xml = Sidereon.CCSDS.OEM.encode(oem, format: :xml)

      # Round-trip through XML
      {:ok, oem2} = Sidereon.CCSDS.OEM.parse(xml)
  """

  alias Sidereon.CCSDS.OEM
  alias Sidereon.NIF

  @typedoc "A Cartesian triple `{x, y, z}`."
  @type vec3 :: {float(), float(), float()}

  @typedoc "Failure reason from the OEM readers."
  @type error :: :missing_field | :invalid_field | :malformed

  defmodule State do
    @moduledoc """
    One Cartesian state sample inside a parsed CCSDS OEM segment.

    `position_km` and `velocity_km_s` are `{x, y, z}` tuples in the segment's
    reference frame. `acceleration_km_s2` is the optional `{x, y, z}`
    acceleration, or `nil` when the message carries no acceleration column.
    """

    @enforce_keys [:epoch, :position_km, :velocity_km_s]
    defstruct [:epoch, :position_km, :velocity_km_s, :acceleration_km_s2]

    @type t :: %__MODULE__{
            epoch: String.t(),
            position_km: OEM.vec3(),
            velocity_km_s: OEM.vec3(),
            acceleration_km_s2: OEM.vec3() | nil
          }
  end

  defmodule Covariance do
    @moduledoc """
    One 6x6 covariance block inside a parsed CCSDS OEM segment.

    `matrix` is a row-major list of six six-element rows. `cov_ref_frame` is the
    optional `COV_REF_FRAME` override, or `nil` to inherit the segment frame.
    """

    @enforce_keys [:epoch, :matrix]
    defstruct [:epoch, :cov_ref_frame, :matrix]

    @type t :: %__MODULE__{
            epoch: String.t(),
            cov_ref_frame: String.t() | nil,
            matrix: [[float()]]
          }
  end

  defmodule Metadata do
    @moduledoc """
    Metadata block for one CCSDS OEM segment.
    """

    @enforce_keys [
      :object_name,
      :object_id,
      :center_name,
      :ref_frame,
      :time_system,
      :start_time,
      :stop_time
    ]
    defstruct [
      :object_name,
      :object_id,
      :center_name,
      :ref_frame,
      :time_system,
      :start_time,
      :stop_time,
      :useable_start_time,
      :useable_stop_time,
      :interpolation,
      :interpolation_degree
    ]

    @type t :: %__MODULE__{
            object_name: String.t(),
            object_id: String.t(),
            center_name: String.t(),
            ref_frame: String.t(),
            time_system: String.t(),
            start_time: String.t(),
            stop_time: String.t(),
            useable_start_time: String.t() | nil,
            useable_stop_time: String.t() | nil,
            interpolation: String.t() | nil,
            interpolation_degree: non_neg_integer() | nil
          }
  end

  defmodule Segment do
    @moduledoc """
    One metadata/data segment of a CCSDS OEM.
    """

    alias Sidereon.CCSDS.OEM.Covariance
    alias Sidereon.CCSDS.OEM.Metadata
    alias Sidereon.CCSDS.OEM.State

    @enforce_keys [:metadata]
    defstruct metadata: nil, states: [], covariances: []

    @type t :: %__MODULE__{
            metadata: Metadata.t(),
            states: [State.t()],
            covariances: [Covariance.t()]
          }
  end

  @enforce_keys [:segments]
  defstruct ccsds_oem_vers: "2.0",
            creation_date: nil,
            originator: nil,
            segments: [],
            skipped_states: 0

  @type t :: %__MODULE__{
          ccsds_oem_vers: String.t(),
          creation_date: String.t() | nil,
          originator: String.t() | nil,
          segments: [Segment.t()],
          skipped_states: non_neg_integer()
        }

  @doc """
  Parse an OEM in either KVN or XML format.

  Format is auto-detected from the first non-whitespace character: `<` routes to
  the XML parser, anything else to the KVN parser.

  Returns `{:ok, %Sidereon.CCSDS.OEM{}}` or `{:error, reason}`.
  """
  @spec parse(String.t()) :: {:ok, t()} | {:error, error()}
  def parse(string) when is_binary(string) do
    if string |> String.trim_leading() |> String.starts_with?("<") do
      parse_xml(string)
    else
      parse_kvn(string)
    end
  end

  @doc """
  Parse an OEM in KVN format explicitly. Skips format auto-detection.
  """
  @spec parse_kvn(String.t()) :: {:ok, t()} | {:error, error()}
  def parse_kvn(text) when is_binary(text) do
    text |> NIF.oem_parse_kvn() |> from_fields()
  end

  @doc """
  Parse an OEM in XML format explicitly. Skips format auto-detection.
  """
  @spec parse_xml(String.t()) :: {:ok, t()} | {:error, error()}
  def parse_xml(text) when is_binary(text) do
    text |> NIF.oem_parse_xml() |> from_fields()
  end

  @doc """
  Encode an OEM.

  ## Options
    * `:format` - `:kvn` (default) or `:xml`
  """
  @spec encode(t(), keyword()) :: String.t()
  def encode(oem, opts \\ [])

  def encode(%__MODULE__{} = oem, opts) do
    case Keyword.get(opts, :format, :kvn) do
      :kvn -> encode_kvn(oem)
      :xml -> encode_xml(oem)
      other -> raise ArgumentError, "unsupported OEM format: #{inspect(other)}"
    end
  end

  @doc """
  Encode an OEM to KVN text explicitly.
  """
  @spec encode_kvn(t()) :: String.t()
  def encode_kvn(%__MODULE__{} = oem), do: NIF.oem_encode_kvn(to_fields(oem))

  @doc """
  Encode an OEM to XML text explicitly.
  """
  @spec encode_xml(t()) :: String.t()
  def encode_xml(%__MODULE__{} = oem), do: NIF.oem_encode_xml(to_fields(oem))

  # --- NIF field marshaling ---

  defp from_fields({:ok, fields}) do
    {:ok,
     %__MODULE__{
       ccsds_oem_vers: fields.ccsds_oem_vers,
       creation_date: fields.creation_date,
       originator: fields.originator,
       segments: Enum.map(fields.segments, &segment_from_fields/1),
       skipped_states: fields.skipped_states
     }}
  end

  defp from_fields({:error, reason}), do: {:error, reason}

  defp segment_from_fields(seg) do
    %Segment{
      metadata: metadata_from_fields(seg.metadata),
      states: Enum.map(seg.states, &state_from_fields/1),
      covariances: Enum.map(seg.covariances, &covariance_from_fields/1)
    }
  end

  defp metadata_from_fields(m) do
    %Metadata{
      object_name: m.object_name,
      object_id: m.object_id,
      center_name: m.center_name,
      ref_frame: m.ref_frame,
      time_system: m.time_system,
      start_time: m.start_time,
      stop_time: m.stop_time,
      useable_start_time: m.useable_start_time,
      useable_stop_time: m.useable_stop_time,
      interpolation: m.interpolation,
      interpolation_degree: m.interpolation_degree
    }
  end

  defp state_from_fields(s) do
    %State{
      epoch: s.epoch,
      position_km: s.position_km,
      velocity_km_s: s.velocity_km_s,
      acceleration_km_s2: s.acceleration_km_s2
    }
  end

  defp covariance_from_fields(c) do
    %Covariance{epoch: c.epoch, cov_ref_frame: c.cov_ref_frame, matrix: c.matrix}
  end

  defp to_fields(%__MODULE__{} = oem) do
    %{
      ccsds_oem_vers: oem.ccsds_oem_vers,
      creation_date: oem.creation_date,
      originator: oem.originator,
      segments: Enum.map(oem.segments, &segment_to_fields/1),
      skipped_states: oem.skipped_states
    }
  end

  defp segment_to_fields(%Segment{} = seg) do
    %{
      metadata: metadata_to_fields(seg.metadata),
      states: Enum.map(seg.states, &state_to_fields/1),
      covariances: Enum.map(seg.covariances, &covariance_to_fields/1)
    }
  end

  defp metadata_to_fields(%Metadata{} = m) do
    %{
      object_name: m.object_name,
      object_id: m.object_id,
      center_name: m.center_name,
      ref_frame: m.ref_frame,
      time_system: m.time_system,
      start_time: m.start_time,
      stop_time: m.stop_time,
      useable_start_time: m.useable_start_time,
      useable_stop_time: m.useable_stop_time,
      interpolation: m.interpolation,
      interpolation_degree: m.interpolation_degree
    }
  end

  defp state_to_fields(%State{} = s) do
    %{
      epoch: s.epoch,
      position_km: s.position_km,
      velocity_km_s: s.velocity_km_s,
      acceleration_km_s2: s.acceleration_km_s2
    }
  end

  defp covariance_to_fields(%Covariance{} = c) do
    %{epoch: c.epoch, cov_ref_frame: c.cov_ref_frame, matrix: c.matrix}
  end
end

File: native/sidereon_nif/src/oem.rs
//! Rustler boundary for the CCSDS OEM KVN and XML reader/writer.
//!
//! Pure glue over `sidereon_core::astro::oem`: decode the raw KVN/XML text or the
//! normalized field map, forward to the crate codec, and encode the unchanged
//! Sidereon result shapes. No grammar, unit handling, or number formatting lives
//! here. Date/time fields cross as raw strings; the Elixir binding owns any
//! resolution to its native `DateTime`. Failure categories cross as atoms.

use rustler::{Encoder, Env, Term};
use sidereon_core::astro::covariance::{Covariance6, Mat6};
use sidereon_core::astro::oem::{
    self as core_oem, Oem, OemCovariance, OemError, OemMetadata, OemSegment, OemState,
};

mod atoms {
    rustler::atoms! {
        ok,
        error,
        missing_field,
        invalid_field,
        malformed
    }
}

/// One Cartesian state sample exchanged with the Elixir binding. Mirrors
/// `%Sidereon.CCSDS.OEM.State{}` with position/velocity as `{x, y, z}` tuples and
/// an optional acceleration tuple.
#[derive(Debug, Clone, rustler::NifMap)]
struct OemStateFields {
    epoch: String,
    position_km: (f64, f64, f64),
    velocity_km_s: (f64, f64, f64),
    acceleration_km_s2: Option<(f64, f64, f64)>,
}

/// One covariance block exchanged with the Elixir binding. The 6x6 matrix is a
/// row-major list of six six-element rows.
#[derive(Debug, Clone, rustler::NifMap)]
struct OemCovarianceFields {
    epoch: String,
    cov_ref_frame: Option<String>,
    matrix: Vec<Vec<f64>>,
}

/// Segment metadata exchanged with the Elixir binding.
#[derive(Debug, Clone, rustler::NifMap)]
struct OemMetadataFields {
    object_name: String,
    object_id: String,
    center_name: String,
    ref_frame: String,
    time_system: String,
    start_time: String,
    stop_time: String,
    useable_start_time: Option<String>,
    useable_stop_time: Option<String>,
    interpolation: Option<String>,
    interpolation_degree: Option<i64>,
}

/// One metadata/data segment exchanged with the Elixir binding.
#[derive(Debug, Clone, rustler::NifMap)]
struct OemSegmentFields {
    metadata: OemMetadataFields,
    states: Vec<OemStateFields>,
    covariances: Vec<OemCovarianceFields>,
}

/// Normalized OEM fields exchanged with the Elixir binding, mirroring
/// `%Sidereon.CCSDS.OEM{}`.
#[derive(Debug, Clone, rustler::NifMap)]
struct OemFields {
    ccsds_oem_vers: String,
    creation_date: Option<String>,
    originator: Option<String>,
    segments: Vec<OemSegmentFields>,
    skipped_states: i64,
}

fn vec3((x, y, z): (f64, f64, f64)) -> [f64; 3] {
    [x, y, z]
}

fn tuple3(v: [f64; 3]) -> (f64, f64, f64) {
    (v[0], v[1], v[2])
}

fn matrix_rows(matrix: &Mat6) -> Vec<Vec<f64>> {
    matrix.iter().map(|row| row.to_vec()).collect()
}

/// Rebuild a 6x6 covariance from the row-major list. Missing or short rows leave
/// the remaining cells at zero; the round-trip carries trusted parsed data, so
/// the matrix is wrapped without re-validation.
fn covariance_from_rows(rows: &[Vec<f64>]) -> Covariance6 {
    let mut matrix: Mat6 = [[0.0_f64; 6]; 6];
    for (out_row, in_row) in matrix.iter_mut().zip(rows) {
        for (slot, value) in out_row.iter_mut().zip(in_row) {
            *slot = *value;
        }
    }
    Covariance6::from_matrix_unchecked(matrix)
}

impl From<OemState> for OemStateFields {
    fn from(s: OemState) -> Self {
        Self {
            epoch: s.epoch,
            position_km: tuple3(s.position_km),
            velocity_km_s: tuple3(s.velocity_km_s),
            acceleration_km_s2: s.acceleration_km_s2.map(tuple3),
        }
    }
}

impl From<OemStateFields> for OemState {
    fn from(f: OemStateFields) -> Self {
        Self {
            epoch: f.epoch,
            position_km: vec3(f.position_km),
            velocity_km_s: vec3(f.velocity_km_s),
            acceleration_km_s2: f.acceleration_km_s2.map(vec3),
        }
    }
}

impl From<OemCovariance> for OemCovarianceFields {
    fn from(c: OemCovariance) -> Self {
        Self {
            epoch: c.epoch,
            cov_ref_frame: c.cov_ref_frame,
            matrix: matrix_rows(c.matrix.as_matrix()),
        }
    }
}

impl From<OemCovarianceFields> for OemCovariance {
    fn from(f: OemCovarianceFields) -> Self {
        Self {
            epoch: f.epoch,
            cov_ref_frame: f.cov_ref_frame,
            matrix: covariance_from_rows(&f.matrix),
        }
    }
}

impl From<OemMetadata> for OemMetadataFields {
    fn from(m: OemMetadata) -> Self {
        Self {
            object_name: m.object_name,
            object_id: m.object_id,
            center_name: m.center_name,
            ref_frame: m.ref_frame,
            time_system: m.time_system,
            start_time: m.start_time,
            stop_time: m.stop_time,
            useable_start_time: m.useable_start_time,
            useable_stop_time: m.useable_stop_time,
            interpolation: m.interpolation,
            interpolation_degree: m.interpolation_degree.map(|d| d as i64),
        }
    }
}

impl From<OemMetadataFields> for OemMetadata {
    fn from(f: OemMetadataFields) -> Self {
        Self {
            object_name: f.object_name,
            object_id: f.object_id,
            center_name: f.center_name,
            ref_frame: f.ref_frame,
            time_system: f.time_system,
            start_time: f.start_time,
            stop_time: f.stop_time,
            useable_start_time: f.useable_start_time,
            useable_stop_time: f.useable_stop_time,
            interpolation: f.interpolation,
            interpolation_degree: f.interpolation_degree.map(|d| d as u32),
        }
    }
}

impl From<OemSegment> for OemSegmentFields {
    fn from(s: OemSegment) -> Self {
        Self {
            metadata: s.metadata.into(),
            states: s.states.into_iter().map(Into::into).collect(),
            covariances: s.covariances.into_iter().map(Into::into).collect(),
        }
    }
}

impl From<OemSegmentFields> for OemSegment {
    fn from(f: OemSegmentFields) -> Self {
        Self {
            metadata: f.metadata.into(),
            states: f.states.into_iter().map(Into::into).collect(),
            covariances: f.covariances.into_iter().map(Into::into).collect(),
        }
    }
}

impl From<Oem> for OemFields {
    fn from(o: Oem) -> Self {
        Self {
            ccsds_oem_vers: o.ccsds_oem_vers,
            creation_date: o.creation_date,
            originator: o.originator,
            segments: o.segments.into_iter().map(Into::into).collect(),
            skipped_states: o.skipped_states as i64,
        }
    }
}

impl From<OemFields> for Oem {
    fn from(f: OemFields) -> Self {
        Self {
            ccsds_oem_vers: f.ccsds_oem_vers,
            creation_date: f.creation_date,
            originator: f.originator,
            segments: f.segments.into_iter().map(Into::into).collect(),
            skipped_states: f.skipped_states.max(0) as usize,
        }
    }
}

/// Map a core OEM failure to its category atom, so the Elixir caller sees a
/// `{:error, atom}` reason rather than a leaked Rust string.
fn error_atom(error: &OemError) -> rustler::Atom {
    match error {
        OemError::MissingField(_) => atoms::missing_field(),
        OemError::InvalidField { .. } => atoms::invalid_field(),
        OemError::Field(_) => atoms::malformed(),
    }
}

fn parse_result<'a>(env: Env<'a>, result: Result<Oem, OemError>) -> Term<'a> {
    match result {
        Ok(parsed) => (atoms::ok(), OemFields::from(parsed)).encode(env),
        Err(e) => (atoms::error(), error_atom(&e)).encode(env),
    }
}

/// Parse a CCSDS OEM in KVN encoding.
#[rustler::nif(schedule = "DirtyCpu")]
fn oem_parse_kvn<'a>(env: Env<'a>, text: String) -> Term<'a> {
    parse_result(env, core_oem::parse_kvn(&text))
}

/// Parse a CCSDS OEM in XML encoding.
#[rustler::nif(schedule = "DirtyCpu")]
fn oem_parse_xml<'a>(env: Env<'a>, text: String) -> Term<'a> {
    parse_result(env, core_oem::parse_xml(&text))
}

/// Serialize normalized OEM fields as CCSDS OEM KVN text.
#[rustler::nif(schedule = "DirtyCpu")]
fn oem_encode_kvn(fields: OemFields) -> String {
    core_oem::encode_kvn(&fields.into())
}

/// Serialize normalized OEM fields as CCSDS OEM XML text.
#[rustler::nif(schedule = "DirtyCpu")]
fn oem_encode_xml(fields: OemFields) -> String {
    core_oem::encode_xml(&fields.into())
}
